### PR TITLE
Jmprieur/easy auth without test project

### PR DESCRIPTION
**Pull request by jmprieur**
## EasyAuth integration

The changes to Microsoft.Identity.Web are the following:

- In Microsoft.Identity.Web we have new classes:
  - `AppServiceAuthenticationBuilderExtensions` declares `service.AddAppServiceAuthentication` which registers the EasyAuth scheme, and the AppServiceAuthenticationHandler. This extension method is called from AddMicrosoftIdentityWebApp. 
  - `AppServiceAuthenticationDefaults`, which defines AppServiceAuthenticationDefaults.AuthenticationScheme (the name of the EasyAuth scheme)
  - `AppServiceAuthenticationHandler` which does the authentication, reading the IdToken and the IdP from the headers, or, when debugging locally from environment variables. It leverages AppServiceAuthenticationInformation static class.
  - `AppServiceAuthenticationInformation` has static properties and methods provide information about EasyAuth based on environment variables set on the AppService machines when EasyAuth is enabled. An important one is "WEBSITE_AUTH_ENABLED", which shows that EasyAuth is enabled when its value is "True". The static property [IsAppServiceAadAuthenticationEnabled]() is used in the code of Microsoft.Identity.Web and Microsoft.Identity.Web.UI that was updated to support EasyAuth. 
  - `AppServicesAuthenticationTokenAcquisition` which is another implementation of ITokenAcquisition, and just reads the token from the header when easy auth is enabled.

- the following classes in Microsoft.Identity.Web were changed to do a different processing when EasyAuth is enabled: MicrosoftIdentityWebAppAuthenticationBuilderExtensions.AddMicrosoftIdentityWebAppInternal. MicrosoftIdentityWebAppAuthenticationBuilder.WebAppCallsWebApiImplementation

-  In Microsoft.Identity.Web.UI, in the AccountController, the SignOut method was changed to redirect to the EasyAuth logout URL (normally **./aad/logout**) in the case EasyAuth is active

### CODE REVIEW

File: src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Microsoft.Identity.Web
+{
+    /// <summary>
+    /// Extension methods related to App Service authentication (Easy Auth).
+    /// </summary>
+    public static class AppServiceAuthenticationBuilderExtensions
+    {
+        /// <summary>
+        /// Add authentication with App services.
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationBuilderExtensions.cs:17]:
```suggestion
        /// Add authentication with App Services.
```

File: src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        /// </summary>
+        /// <param name="builder">Authentication build.</param>
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationBuilderExtensions.cs:19]:
```suggestion
        /// <param name="builder">Authentication builder.</param>
```

File: src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationDefaults.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.Identity.Web
+{
+    /// <summary>
+    /// Default values related to AppServiceAuthentication authentication handler.
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationDefaults.cs:7]:
```suggestion
    /// Default values related to AppServiceAuthentication handler.
```

File: src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationDefaults.cs [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    /// </summary>
+    public class AppServiceAuthenticationDefaults
+    {
+        /// <summary>
+        /// The default value used for AppServiceAuthenticationOptions.AuthenticationScheme.
+        /// </summary>
+        public const string AuthenticationScheme = "EasyAuth";
+    }
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationDefaults.cs:15]: nit: "EasyAuth" or "AppServices"? what do people use?

File: src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationHandler.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                string? idp = GetIdp();
+
+                if (idToken != null && idp != null)
+                {
+                    JsonWebToken jsonWebToken = new JsonWebToken(idToken);
+                    ClaimsPrincipal claimsPrincipal = new ClaimsPrincipal(new ClaimsIdentity(
+                        jsonWebToken.Claims,
+                        idp,
+                        "name", // v1.0
+                        ClaimsIdentity.DefaultRoleClaimType));
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationHandler.cs:57]: we have a Constants.NameClaim you can use here.

File: src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationHandler.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+              ILoggerFactory logger,
+              UrlEncoder encoder,
+              ISystemClock clock)
+              : base(options, logger, encoder, clock)
+        {
+        }
+
+        // Constants
+        private const string EasyAuthIdTokenHeader = "X-MS-TOKEN-AAD-ID-TOKEN";
+        private const string EasyAuthIdpTokenHeader = "X-MS-CLIENT-PRINCIPAL-IDP";
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationHandler.cs:40]: nit: maybe move this to the constants class?

File: src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationHandler.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            }
+
+            // Try another handler
+            return Task.FromResult(AuthenticateResult.NoResult());
+        }
+
+        private string? GetIdp()
+        {
+            string? idp = Context.Request.Headers[EasyAuthIdpTokenHeader];
+#if DEBUG
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationHandler.cs:72]: could we ever have any null refs on any of these?

File: src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationInformation.cs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+using System.Linq;
+
+namespace Microsoft.Identity.Web
+{
+    /// <summary>
+    /// Information about the AppService configuration on the host.
+    /// </summary>
+    public static class AppServiceAuthenticationInformation
+    {
+        // Environment variables.
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationInformation.cs:15]: i think we could move all of these to the constant file. but just my 2 cents.

File: src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationInformation.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        /// Is AppService authentication enabled?.
+        /// </summary>
+        public static bool IsAppServiceAadAuthenticationEnabled
+        {
+            get
+            {
+                return (Environment.GetEnvironmentVariable(EasyAuthEnabledEnvironmentVariable) == "True")
+                    && Environment.GetEnvironmentVariable(EasyAuthIdentityProviderEnvironmentVariable) == EasyAuthAzureActiveDirectory;
+            }
+        }
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationInformation.cs:37]: as we discussed yesterday, will there be an error w/aka.ms link for customers using FB or another IDP that's not AAD?

File: src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationInformation.cs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        internal static string? ClientId
+        {
+            get
+            {
+                return Environment.GetEnvironmentVariable(EasyAuthClientIdEnvironmentVariable);
+            }
+        }
+
+        /// <summary>
+        /// ClientID of the Easy Auth web site.
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationInformation.cs:62]:
```suggestion
        /// Client secret of the Easy Auth web site.
```

File: src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationInformation.cs [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        internal static string? ClientSecret
+        {
+            get
+            {
+                return Environment.GetEnvironmentVariable(EasyAuthClientSecretEnvironmentVariable);
+            }
+        }
+
+        /// <summary>
+        /// ClientID of the Easy Auth web site.
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationInformation.cs:73]:
```suggestion
        /// Issuer of the Easy Auth web site.
```

**Review by jennyf19 — APPROVED** [review]
:shipit:

File: src/Microsoft.Identity.Web/Microsoft.Identity.Web.xml [hunk 12]
```diff
@@ -778,6 +778,111 @@
...
         </member>
+        <member name="T:Microsoft.Identity.Web.AppServiceAuthenticationBuilderExtensions">
+            <summary>
+            Extension methods related to App Service authentication (Easy Auth).
+            </summary>
+        </member>
+        <member name="M:Microsoft.Identity.Web.AppServiceAuthenticationBuilderExtensions.AddAppServiceAuthentication(Microsoft.AspNetCore.Authentication.AuthenticationBuilder)">
+            <summary>
+            Add authentication with App services.
+            </summary>
```
[the remark this answers is not in the document]
Comment (jmprieur) [reply src/Microsoft.Identity.Web/Microsoft.Identity.Web.xml:789]: AppServicesAuthNZ ?

[the remark this answers is not in the document]
Comment (jennyf19) [reply src/Microsoft.Identity.Web/Microsoft.Identity.Web.xml:789]: Or just AppServicesAuth

**Comment by jennyf19** [comment]
[PR](https://github.com/AzureAD/microsoft-identity-web/pull/701) addressing the naming things

File: tests/WebAppCallsMicrosoftGraph/Pages/Debugging.cshtml [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    string[] importantHeaders = new string[] { "X-MS-TOKEN-AAD-ID-TOKEN", "X-MS-CLIENT-PRINCIPAL-IDP", "X-MS-TOKEN-AAD-ACCESS-TOKEN" };
+}
+
+
+<h1>Debug</h1>
+
+To debug locally the application as if it was deployed in App services:
+<ol>
+    <li>
+        Run the application with Easy auth as deployed, but at the baseUrl/Debugging address (for instance: <a href='https://mdidwebtestforeasyauthintegration.azurewebsites.net/Debugging'>https://mdidwebtestforeasyauthintegration.azurewebsites.net/Debugging</a>)
```
Comment (jennyf19) [inline tests/WebAppCallsMicrosoftGraph/Pages/Debugging.cshtml:17]:
```suggestion
        Run the application with App Services authentication as deployed, but at the baseUrl/Debugging address (for instance: <a href='https://mdidwebtestforeasyauthintegration.azurewebsites.net/Debugging'>https://mdidwebtestforeasyauthintegration.azurewebsites.net/Debugging</a>)
```

File: tests/WebAppCallsMicrosoftGraph/Pages/Debugging.cshtml [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    </li>
+    <li>
+        Copy the following JSon blob into the <b>lauchsettings.json</b> for the profile of your choice, in the <b>environmentVariables</b> section, paste the following:
+    </li>
+
+    <li>
+        Run Visual Studio with debugging with the profile for which you have added the environement variables.
```
Comment (jennyf19) [inline tests/WebAppCallsMicrosoftGraph/Pages/Debugging.cshtml:24]:
```suggestion
        Run Visual Studio with debugging with the profile for which you have added the environment variables.
```

File: src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationInformation.cs (hunk 9 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web/EasyAuth/AppServiceAuthenticationInformation.cs]: I added an issue to do it later: https://github.com/AzureAD/microsoft-identity-web/issues/702

**Pull request merged**